src/node_buffer.cc

```diff
@@ -113,19 +164,43 @@ Local<Object> New(Isolate* isolate, size_t length) {
 }
 
 
-// TODO(trevnorris): these have a flaw by needing to call the Buffer inst then
-// Alloc. continue to look for a better architecture.
 Local<Object> New(Environment* env, size_t length) {
   EscapableHandleScope scope(env->isolate());
 
-  CHECK_LE(length, kMaxLength);
+  if (using_old_buffer) {
+    CHECK_LE(length, kMaxLength);
 
-  Local<Value> arg = Uint32::NewFromUnsigned(env->isolate(), length);
-  Local<Object> obj = env->buffer_constructor_function()->NewInstance(1, &arg);
+    Local<Value> arg = Uint32::NewFromUnsigned(env->isolate(), length);
+    Local<Object> obj =
+        env->buffer_constructor_function()->NewInstance(1, &arg);
 
-  smalloc::Alloc(env, obj, length);
+    smalloc::Alloc(env, obj, length);
 
-  return scope.Escape(obj);
+    return scope.Escape(obj);
+  }
+
+  if (!v8::internal::Internals::IsValidSmi(length)) {
```

Comment:
I suggest you add a comment here explaining the reason for using something from the v8::internal namespace. :-)